contracts/gateway/OSTPrime.sol

```diff
@@ -104,65 +110,109 @@ contract OSTPrime is UtilityToken, OSTPrimeConfig {
         payable
+        returns (bool success_)
     {
-        require(msg.value == TOKENS_MAX);
+        require(
+            initialized == false,
+            "Contract is already initialized."
+        );
+
+        require(
+            msg.value == TOKENS_MAX,
+            "Payable amount must be equal to total supply of token."
+        );
+
         initialized = true;
+
+        success_ = true;
     }
 
+
+    /* External functions. */
+
     /**
-     * @notice convert the OST utility token to base token
+     * @notice convert the OST Prime token to OST Prime base token.
      *
-     * @param _amount Amount of basetoken
+     * @param _amount Amount of OST Prime token to convert to base token.
      *
-     * @return `true` if claim was successfully progressed
+     * @return success_ `true` if unwrap was successful.
      */
-    function claim(
+    function unwrap(
         uint256 _amount
     )
-        public
+        external
         onlyInitialized
-        returns (bool /* success */)
+        returns (bool success_)
     {
+        require(
+            _amount > 0,
+            "Amount should not be zero."
+        );
+
         require(
             _amount <= balances[msg.sender],
-            "Insufficient balance"
+            "Insufficient balance."
+        );
+
+        require(
+            address(this).balance >= _amount,
+            "Contact balance should not be less than the unwrap amount."
         );
 
-        assert(address(this).balance >= _amount);
 
-        transfer(address(this),_amount);
+        balances[msg.sender] = balances[msg.sender].sub(_amount);
+        balances[address(this)] = balances[address(this)].add(_amount);
         msg.sender.transfer(_amount);
 
-        emit Claim(msg.sender, _amount, totalTokenSupply, address(this));
+        emit Transfer(msg.sender, address(this), _amount);
+        emit TokenUnwrapped(msg.sender, _amount);
 
-        return true;
+        success_ = true;
     }
 
     /**
-     * @notice convert the base token to OST utility token
+     * @notice convert OST Prime base token to OST Prime token.
      *
-     * @param _amount Amount of utility token
-     *
-     * @return `true` if claim was successfully progressed
+     * @return success_ `true` if claim was successfully progressed.
      */
-    function redeem(uint256 _amount)
-        public
+    function wrap()
+        external
         onlyInitialized
         payable
-        returns (bool /** success */)
+        returns (bool success_)
     {
-        require(msg.value == _amount);
-        assert(address(this).balance >= _amount);
+        uint256 amount = msg.value;
+        address account = msg.sender;
 
-        allowed[address(this)][msg.sender] = _amount;
-        transferFrom(address(this), msg.sender, _amount);
+        require(
+            amount > 0,
+            "Payable amount should not be zero."
+        );
+
+        require(
+            address(account).balance >= amount,
```

Comment:
I didn't understand this part. 
Theoretically, account eth balance should be reduced when this transaction is triggered. By this logic, the caller can't wrap all the amount. 

A unit test case to wrap the total balance amount should validate this. 

@schemar  @gulshanvasnani  What do you think?